.template.Debian/docker-entrypoint.sh

```diff
@@ -238,23 +236,31 @@ docker_setup_db() {
 	fi
 }
 
+_mysql_passfile() {
```

Comment:
Running this every time we execute an SQL seems a bit odd, especially since it leads to the side effect of having to manipulate the user-set MYSQL_ROOT_PASSWORD variable in various places around the script.

Reply:
The intent behind this is to keep the password always in-memory only instead of committing it to disk (and then also having a file we need to clean up).

Reply:
Right, but right now it adds confusing complexity to the script, since we have to worry about the state of MYSQL_ROOT_PASSWORD in multiple places where its value should not be relevant. Maybe just add a simple flag that's set once the root password is set in the db, and have _mysql_passfile check it?

Reply:
I see two interpretations.  We could make `_mysql_passfile` take an argument, but giving an argument to `docker_process_sql` to then pass along is a bit more complex since it currently passes everything directly to `mysql`. Since `mysql` allows a bareword of the database to connect to, any keyword we pick couldn't be used as a database name to connect to or it would have to be a flag that would never be used by `mysql`.

The other solution would be another global value, like `MYSQL_ROOT_PASSWORD_SET`, but the wrong name and users might try to set it themselves and assume root user creation would be skipped (or some other interpretation).

Reply:
Here's my rough version using a flag. I also added `DATABASE_ALREADY_EXISTS` so that `docker_temp_server_start` could be used on an already initialized database (e.x. for an `always-initdb.d` dir).

```diff
diff --git a/.template.Debian/docker-entrypoint.sh b/.template.Debian/docker-entrypoint.sh
index 9b5ce17..85f8ec8 100755
--- a/.template.Debian/docker-entrypoint.sh
+++ b/.template.Debian/docker-entrypoint.sh
@@ -95,9 +95,13 @@ docker_temp_server_start() {
 		mysql_note "Waiting for server startup"
 		local i
 		for i in {30..0}; do
-			# unset MYSQL_ROOT_PASSWORD for just docker_process_sql
+			# only use the root password if the database has already been initializaed
 			# so that it won't try to fill in a password file when it hasn't been set yet
-			if MYSQL_ROOT_PASSWORD= docker_process_sql --database=mysql <<<'SELECT 1' &> /dev/null; then
+			extraArgs=()
+			if [ -z "$DATABASE_ALREADY_EXISTS" ]; then
+				extraArgs+=( '--dont-use-mysql-root-password' )
+			fi
+			if docker_process_sql "${extraArgs[@]}" --database=mysql <<<'SELECT 1' &> /dev/null; then
 				break
 			fi
 			sleep 1
@@ -168,19 +172,29 @@ docker_setup_env() {
 	file_env 'MYSQL_USER'
 	file_env 'MYSQL_PASSWORD'
 	file_env 'MYSQL_ROOT_PASSWORD'
+
+	declare -g DATABASE_ALREADY_EXISTS
+	if [ -d "$DATADIR/mysql" ]; then
+		DATABASE_ALREADY_EXISTS='true'
+	fi
 }
 
 # Execute sql script, passed via stdin
-# usage: docker_process_sql [mysql-cli-args]
+# usage: docker_process_sql [--dont-use-mysql-root-password] [mysql-cli-args]
 #    ie: docker_process_sql --database=mydb <<<'INSERT ...'
-#    ie: docker_process_sql --database=mydb <my-file.sql
+#    ie: docker_process_sql --dont-use-mysql-root-password --database=mydb <my-file.sql
 docker_process_sql() {
+	passfileArgs=()
+	if [ '--dont-use-mysql-root-password' = "$1" ]; then
+		passfileArgs+=( "$1" )
+		shift
+	fi
 	# args sent in can override this db, since they will be later in the command
 	if [ -n "$MYSQL_DATABASE" ]; then
 		set -- --database="$MYSQL_DATABASE" "$@"
 	fi
 
-	mysql --defaults-file=<( _mysql_passfile ) --protocol=socket -uroot -hlocalhost --socket="${SOCKET}" "$@"
+	mysql --defaults-file=<( _mysql_passfile "${passfileArgs[@]}") --protocol=socket -uroot -hlocalhost --socket="${SOCKET}" "$@"
 }
 
 # Initializes database with timezone info and root password, plus optional extra db/user
@@ -190,9 +204,8 @@ docker_setup_db() {
 		# sed is for https://bugs.mysql.com/bug.php?id=20545
 		mysql_tzinfo_to_sql /usr/share/zoneinfo \
 			| sed 's/Local time zone must be set--see zic manual page/FCTY/' \
-			| MYSQL_ROOT_PASSWORD= docker_process_sql --database=mysql
-		# unset MYSQL_ROOT_PASSWORD for just docker_process_sql
-		# so that it won't try to fill in a password file when it hasn't been set yet
+			| docker_process_sql --dont-use-mysql-root-password --database=mysql
+			# tell docker_process_sql to not use MYSQL_ROOT_PASSWORD since it is not set yet
 	fi
 	# Generate random root password
 	if [ -n "$MYSQL_RANDOM_ROOT_PASSWORD" ]; then
@@ -211,9 +224,8 @@ docker_setup_db() {
 		EOSQL
 	fi
 
-	# unset MYSQL_ROOT_PASSWORD for just docker_process_sql
-	# so that it won't try to fill in a password file when it is just now being set
-	MYSQL_ROOT_PASSWORD= docker_process_sql --database=mysql <<-EOSQL
+	# tell docker_process_sql to not use MYSQL_ROOT_PASSWORD since it is just now being set
+	docker_process_sql --dont-use-mysql-root-password --database=mysql <<-EOSQL
 		-- What's done in this file shouldn't be replicated
 		--  or products like mysql-fabric won't work
 		SET @@SESSION.SQL_LOG_BIN=0;
@@ -248,7 +260,7 @@ _mysql_passfile() {
 	# echo the password to the "file" the client uses
 	# the client command will use process substitution to create a file on the fly
 	# ie: --defaults-file=<( _mysql_passfile )
-	if [ -n "$MYSQL_ROOT_PASSWORD" ]; then
+	if [ '--dont-use-mysql-root-password' != "$1" ] && [ -n "$MYSQL_ROOT_PASSWORD" ]; then
 		cat <<-EOF
 			[client]
 			password="${MYSQL_ROOT_PASSWORD}"
@@ -301,8 +313,8 @@ _main() {
 			exec gosu mysql "$BASH_SOURCE" "$@"
 		fi
 
-		# If this is true then there's no database, and it needs to be initialized
-		if [ ! -d "$DATADIR/mysql" ]; then
+		# there's no database, so it needs to be initialized
+		if [ -z "$DATABASE_ALREADY_EXISTS" ]; then
 			docker_verify_minimum_env
 			docker_init_database_dir "$@"
```

Reply:
Yeah, I think this could work (I still have some niggling doubts, but don't see any obvious better ways to handle it, and would rather not have a 2nd anniversary for this thing).